draft-irtf-cfrg-opaque.md

```diff
@@ -385,145 +385,329 @@ variant) of the OPRF described in {{I-D.irtf-cfrg-voprf}}.
-Clients MUST NOT use the same key pair (client_private_key, client_public_key) for two different accounts.
+OPAQUE makes use of a structure `Envelope` to manage client credentials. This envelope holds information about its format and content for the client to obtain its authentication material.
 
-To begin, C chooses its password, and S chooses its own pair of private-public
-AKE keys (server_private_key, server_public_key) for use with the AKE. S can use
-the same pair of keys with multiple clients. These steps can happen offline, i.e.,
-before the registration phase. Once complete, the registration process proceeds as follows:
+OPAQUE allows applications to either provide custom client private and public keys for authentication, or to generate them internally, making the application oblivious to the client's private key.
 
+These two options are defined as the `Internal Key` and `External Key` modes.  See {{envelope-modes}} for their specifications.
+
+Credential material is used to pin key material to identities if desired. If no identity is given for a party, its value MUST default to its public key. The following types of credentials are hereby considered:
+
+- client_private_key: The encoded client private key for the AKE protocol.
+- server_public_key: The encoded server public key for the AKE protocol.
+- client_identity: The client identity. This is an application-specific value, e.g., an e-mail address or normal account name. If not specified, it defaults to the client's public key.
+- server_identity: The server identity. This is typically a domain name, e.g., example.com.  If not specified, it defaults to the server's public key.
+  See {{identities}} for information about this identity.
+
+They are used in the `SecretCredentials` and `CleartextCredentials` structures are defined as follows:
 ~~~
- Client (password, creds)            Server (server_private_key, server_public_key)
- --------------------------------------------------------------------
- (request, blind) = CreateRegistrationRequest(password)
+struct {
+ opaque client_private_key[Nsk];
+} SecretCredentials;
 
-                               request
-                      ------------------------->
+struct {
+ opaque server_public_key[Npk];
+ opaque client_identity<0..2^16-1>;
+ opaque server_identity<0..2^16-1>;
+} CleartextCredentials;
+~~~
 
-            (response, oprf_key) = CreateRegistrationResponse(request, server_public_key)
+During protocol execution, they are recommended to be stored in a `Credentials` object with the following named fields:
 
-                               response
-                      <-------------------------
+- `client_private_key`, the client's private key
+- `client_public_key`, the client's public key corresponding to `client_private_key`
+- `client_identity`, an optional client identity (present only in the `custom_identifier` mode)
+- `server_identity`, an optional server identity (present only in the `custom_identifier` mode)
```

Comment:
reference to `custom_identifier` mode which needs to be updated

Reply:
That's a leftover, well spotted!